# **Bioinformatics Project - Computational Drug Discovery [Part 3] Descriptor Calculation and Dataset Preparation(Modified Version towards unknown smiles predcition purpose)**

Chanin Nantasenamat

[*'Data Professor' YouTube channel*](http://youtube.com/dataprofessor)

Modified by quantaosun@gmail.com


In **Part 3**, we will be calculating molecular descriptors that are essentially quantitative description of the compounds in the dataset. Finally, we will be preparing this into a dataset for subsequent model prediction in notebook 5.

---

## **Download PaDEL-Descriptor**

In [1]:
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.sh

--2022-01-09 12:11:01--  https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip [following]
--2022-01-09 12:11:01--  https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25768637 (25M) [application/zip]
Saving to: ‘padel.zip’

padel.zip           100%[===================>]  24.57M   163MB/s    in 0.2s    

2022-01-09 12:11:02 (163 MB/s) - ‘padel.zip’ saved [25768637/25768637]

--2022-01-09 12:11:02--  https://github

In [2]:
! unzip padel.zip

Archive:  padel.zip
   creating: PaDEL-Descriptor/
  inflating: __MACOSX/._PaDEL-Descriptor  
  inflating: PaDEL-Descriptor/MACCSFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._MACCSFingerprinter.xml  
  inflating: PaDEL-Descriptor/AtomPairs2DFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._AtomPairs2DFingerprinter.xml  
  inflating: PaDEL-Descriptor/EStateFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._EStateFingerprinter.xml  
  inflating: PaDEL-Descriptor/Fingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._Fingerprinter.xml  
  inflating: PaDEL-Descriptor/.DS_Store  
  inflating: __MACOSX/PaDEL-Descriptor/._.DS_Store  
   creating: PaDEL-Descriptor/license/
  inflating: __MACOSX/PaDEL-Descriptor/._license  
  inflating: PaDEL-Descriptor/KlekotaRothFingerprintCount.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._KlekotaRothFingerprintCount.xml  
  inflating: PaDEL-Descriptor/config  
  inflating: __MACOSX/PaDEL-Descriptor/._config  
  inf

## **Load unknown smiles data, called "example.txt"**

Download the curated ChEMBL bioactivity data that has been pre-processed from Parts 1 and 2 of this Bioinformatics Project series. Here we will be using the **bioactivity_data_3class_pIC50.csv** file that essentially contain the pIC50 values that we will be using for building a regression model.

In [3]:
import pandas as pd

# We will convert a raw txt file with two lines of smiles to a standard dataframe by a csv intermediate

In [4]:
df3 = pd.read_csv('/content/sars-cov-2_bioactivity_data_processed_2classes_pIC50.csv')

In [5]:
df3

Unnamed: 0 molecule_chembl_id  ... bioactivity_class     pIC50
0            10         CHEMBL1751  ...            active  6.721246
1            11      CHEMBL1526240  ...          inactive  4.873869
2            17          CHEMBL496  ...            active  6.045757
3            28         CHEMBL1448  ...            active  6.552842
4            34         CHEMBL1242  ...          inactive  4.552842
..          ...                ...  ...               ...       ...
134         278      CHEMBL3913746  ...            active  6.397940
135         279      CHEMBL1254577  ...          inactive  4.934047
136         282      CHEMBL3967196  ...          inactive  4.706637
137         283       CHEMBL601661  ...          inactive  4.665144
138         284       CHEMBL178334  ...          inactive  4.524765

[139 rows x 5 columns]

Next the txt file will be save as a csv file, then import back again, to beautify the structure.

In [7]:
selection=['canonical_smiles']
df3[selection]

canonical_smiles
0    C[C@H]1O[C@@H](O[C@H]2[C@@H](O)C[C@H](O[C@H]3[...
1    CC(C)=CCCC1(C)C=Cc2c(O)c3c(c(CC=C(C)C)c2O1)OC1...
2          Oc1c(Cl)cc(Cl)c(Cl)c1Cc1c(O)c(Cl)cc(Cl)c1Cl
3          O=C(Nc1ccc([N+](=O)[O-])cc1Cl)c1cc(Cl)ccc1O
4                          Nc1ccc(/N=N/c2ccccc2)c(N)n1
..                                                 ...
134  CC(C)(O)CNc1ncc(-c2ccnc(Nc3ccnc(Cl)c3)n2)c(CC2...
135  O=C(NCCN1CCC2(CC1)C(=O)NCN2c1cccc(F)c1)c1ccc2c...
136        CCCCn1c(NC(=O)c2cccs2)c(C#N)c2nc3ccccc3nc21
137  CNC(=O)Nc1ccc(-c2nc(N3CC4CCC(C3)O4)c3cnn(C4CCC...
138  CC(C)N1CCN(c2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O...

[139 rows x 1 columns]

In [8]:
df4_selection = df3[selection]
df4_selection

canonical_smiles
0    C[C@H]1O[C@@H](O[C@H]2[C@@H](O)C[C@H](O[C@H]3[...
1    CC(C)=CCCC1(C)C=Cc2c(O)c3c(c(CC=C(C)C)c2O1)OC1...
2          Oc1c(Cl)cc(Cl)c(Cl)c1Cc1c(O)c(Cl)cc(Cl)c1Cl
3          O=C(Nc1ccc([N+](=O)[O-])cc1Cl)c1cc(Cl)ccc1O
4                          Nc1ccc(/N=N/c2ccccc2)c(N)n1
..                                                 ...
134  CC(C)(O)CNc1ncc(-c2ccnc(Nc3ccnc(Cl)c3)n2)c(CC2...
135  O=C(NCCN1CCC2(CC1)C(=O)NCN2c1cccc(F)c1)c1ccc2c...
136        CCCCn1c(NC(=O)c2cccs2)c(C#N)c2nc3ccccc3nc21
137  CNC(=O)Nc1ccc(-c2nc(N3CC4CCC(C3)O4)c3cnn(C4CCC...
138  CC(C)N1CCN(c2ccc(C(=O)c3c(-c4ccc(O)cc4)sc4cc(O...

[139 rows x 1 columns]

In [9]:
df4_selection.to_csv('molecule.smi', sep= '\t', index=False, header=False)

In [10]:
! cat molecule.smi | head -5

C[C@H]1O[C@@H](O[C@H]2[C@@H](O)C[C@H](O[C@H]3[C@@H](O)C[C@H](O[C@H]4CC[C@]5(C)[C@H]6C[C@@H](O)[C@]7(C)[C@@H](C8=CC(=O)OC8)CC[C@]7(O)[C@@H]6CC[C@@H]5C4)O[C@@H]3C)O[C@@H]2C)C[C@H](O)[C@@H]1O
CC(C)=CCCC1(C)C=Cc2c(O)c3c(c(CC=C(C)C)c2O1)OC12C(=CC4CC1C(C)(C)OC2(C/C=C(/C)C(=O)O)C4O)C3=O
Oc1c(Cl)cc(Cl)c(Cl)c1Cc1c(O)c(Cl)cc(Cl)c1Cl
O=C(Nc1ccc([N+](=O)[O-])cc1Cl)c1cc(Cl)ccc1O
Nc1ccc(/N=N/c2ccccc2)c(N)n1


In [11]:
! cat molecule.smi | wc -l

139


## **Calculate fingerprint descriptors, for the unknown smiles**


### **Calculate PaDEL descriptors**

In [12]:
! cat padel.sh

java -Xms1G -Xmx1G -Djava.awt.headless=true -jar ./PaDEL-Descriptor/PaDEL-Descriptor.jar -removesalt -standardizenitro -fingerprints -descriptortypes ./PaDEL-Descriptor/PubchemFingerprinter.xml -dir ./ -file descriptors_output.csv


In [13]:
! bash padel.sh

Processing AUTOGEN_molecule_1 in molecule.smi (1/139). 
Processing AUTOGEN_molecule_2 in molecule.smi (2/139). 
Processing AUTOGEN_molecule_3 in molecule.smi (3/139). Average speed: 5.57 s/mol.
Processing AUTOGEN_molecule_4 in molecule.smi (4/139). Average speed: 3.22 s/mol.
Processing AUTOGEN_molecule_5 in molecule.smi (5/139). Average speed: 2.32 s/mol.
Processing AUTOGEN_molecule_6 in molecule.smi (6/139). Average speed: 1.78 s/mol.
Processing AUTOGEN_molecule_7 in molecule.smi (7/139). Average speed: 1.47 s/mol.
Processing AUTOGEN_molecule_8 in molecule.smi (8/139). Average speed: 1.35 s/mol.
Processing AUTOGEN_molecule_9 in molecule.smi (9/139). Average speed: 1.26 s/mol.
Processing AUTOGEN_molecule_10 in molecule.smi (10/139). Average speed: 1.12 s/mol.
Processing AUTOGEN_molecule_11 in molecule.smi (11/139). Average speed: 1.01 s/mol.
Processing AUTOGEN_molecule_12 in molecule.smi (12/139). Average speed: 0.97 s/mol.
Processing AUTOGEN_molecule_13 in molecule.smi (13/139). Avera

In [14]:
! ls -l

total 25468
-rw-r--r-- 1 root root   259355 Jan  9 12:14 descriptors_output.csv
drwxr-xr-x 3 root root     4096 Jan  9 12:11 __MACOSX
-rw-r--r-- 1 root root     8735 Jan  9 12:13 molecule.smi
drwxrwxr-x 4 root root     4096 May 30  2020 PaDEL-Descriptor
-rw-r--r-- 1 root root      231 Jan  9 12:11 padel.sh
-rw-r--r-- 1 root root 25768637 Jan  9 12:11 padel.zip
drwxr-xr-x 1 root root     4096 Dec 23 14:32 sample_data
-rw-r--r-- 1 root root    14767 Jan  9 12:11 sars-cov-2_bioactivity_data_processed_2classes_pIC50.csv


## **Preparing the X matrices**

### **X data matrix**

In [15]:
df3_X = pd.read_csv('descriptors_output.csv')

In [16]:
df3_X

Name  PubchemFP0  ...  PubchemFP879  PubchemFP880
0      AUTOGEN_molecule_2           1  ...             0             0
1      AUTOGEN_molecule_3           1  ...             0             0
2      AUTOGEN_molecule_4           1  ...             0             0
3      AUTOGEN_molecule_1           1  ...             0             0
4      AUTOGEN_molecule_5           1  ...             0             0
..                    ...         ...  ...           ...           ...
134  AUTOGEN_molecule_134           1  ...             0             0
135  AUTOGEN_molecule_137           1  ...             0             0
136  AUTOGEN_molecule_136           1  ...             0             0
137  AUTOGEN_molecule_138           1  ...             0             0
138  AUTOGEN_molecule_139           1  ...             0             0

[139 rows x 882 columns]

In [17]:
df3_X = df3_X.drop(columns=['Name'])
df3_X

PubchemFP0  PubchemFP1  ...  PubchemFP879  PubchemFP880
0             1           1  ...             0             0
1             1           0  ...             0             0
2             1           1  ...             0             0
3             1           1  ...             0             0
4             1           1  ...             0             0
..          ...         ...  ...           ...           ...
134           1           1  ...             0             0
135           1           1  ...             0             0
136           1           1  ...             0             0
137           1           1  ...             0             0
138           1           1  ...             0             0

[139 rows x 881 columns]

 # Y variable

In [21]:
df3_Y = df3['pIC50']
df3_Y

0      6.721246
1      4.873869
2      6.045757
3      6.552842
4      4.552842
         ...   
134    6.397940
135    4.934047
136    4.706637
137    4.665144
138    4.524765
Name: pIC50, Length: 139, dtype: float64

# Combine X and Y

In [22]:
dataset3 = pd.concat([df3_X,df3_Y], axis=1)
dataset3

PubchemFP0  PubchemFP1  PubchemFP2  ...  PubchemFP879  PubchemFP880     pIC50
0             1           1           1  ...             0             0  6.721246
1             1           0           0  ...             0             0  4.873869
2             1           1           0  ...             0             0  6.045757
3             1           1           1  ...             0             0  6.552842
4             1           1           0  ...             0             0  4.552842
..          ...         ...         ...  ...           ...           ...       ...
134           1           1           1  ...             0             0  6.397940
135           1           1           1  ...             0             0  4.934047
136           1           1           1  ...             0             0  4.706637
137           1           1           1  ...             0             0  4.665144
138           1           1           1  ...             0             0  4.524765

[139 rows x 882 columns]

# You may have noticed that we now have a very long bit of 882 descriptor set, but when we built the model earlier, there are much less descriptors. This mismatch will be addressed later.

In [24]:
dataset3.to_csv('sars-cov-2_bioactivity_data_2class_pIC50_pubchem_fp.csv', index=False)

# **Let's download the CSV file to your local computer for the Part 3B (Model Building).**